src/Lib.P2P/Cryptography/CtrStreamCipher.cs

```diff
@@ -0,0 +1,122 @@
+            counterOut = new byte[blockSize];
+            IV = new byte[blockSize];
+        }
+
+        /// <summary>
+        ///   The name of this algorithm.
+        /// </summary>
+        public string AlgorithmName => cipher.AlgorithmName + "/CTR";
+
+        /// <summary>
+        ///   Init the cipher.
+        /// </summary>
+        /// <param name="forEncryption">
+        ///   Ignored.
+        /// </param>
+        /// <param name="parameters">
+        ///   Must be a <see cref="ParametersWithIV"/>.
+        /// </param>
+        /// <example>
+        /// var encrypt = new CtrStreamCipher(new AesEngine());
+        /// var p = new ParametersWithIV(new KeyParameter(key), iv);
+        /// encrypt.Init(true, p);
+        /// </example>
+        public void Init(bool forEncryption, ICipherParameters parameters)
+        {
+            var ivParam = parameters as ParametersWithIV;
+            if (ivParam == null)
+                throw new ArgumentException("CTR mode requires ParametersWithIV", "parameters");
+
+            IV = Arrays.Clone(ivParam.GetIV());
+
+            if (blockSize < IV.Length)
+                throw new ArgumentException("CTR mode requires IV no greater than: " + blockSize + " bytes.");
+
+            var maxCounterSize = Math.Min(8, blockSize / 2);
+            if (blockSize - IV.Length > maxCounterSize)
+                throw new ArgumentException("CTR mode requires IV of at least: " + (blockSize - maxCounterSize) +
+                    " bytes.");
+
+            // if null it's an IV changed only.
+            if (ivParam.Parameters != null) cipher.Init(true, ivParam.Parameters);
+
+            Reset();
+        }
+
+        /// <inheritdoc />
+        public void Reset()
+        {
+            byteCount = 0;
+            Arrays.Fill(counter, (byte) 0);
+            Array.Copy(IV, 0, counter, 0, IV.Length);
+            cipher.Reset();
+        }
+
+        /// <inheritdoc />
+        public void ProcessBytes(byte[] input, int inOff, int length, byte[] output, int outOff)
+        {
+            if (outOff + length > output.Length) throw new DataLengthException("Output buffer too short");
+
+            if (inOff + length > input.Length) throw new DataLengthException("Input buffer too small");
+
+            var inStart = inOff;
+            var inEnd = inOff + length;
+            var outStart = outOff;
+
+            while (inStart < inEnd) output[outStart++] = ReturnByte(input[inStart++]);
+        }
+
+        /// <inheritdoc />
+        public byte ReturnByte(byte input)
+        {
+            if (byteCount == 0)
+            {
+                cipher.ProcessBlock(counter, 0, counterOut, 0);
+
+                // Increment the counter
+                var j = counter.Length;
+                while (--j >= 0 && ++counter[j] == 0) { }
+            }
+
+            var rv = (byte) (counterOut[byteCount++] ^ input);
+            if (byteCount == counterOut.Length) byteCount = 0;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)